api/features/feature_segments/views.py

```diff
@@ -60,6 +65,38 @@ def get_serializer(self, *args, **kwargs):
             kwargs = {**kwargs, "many": True, "partial": True}
         return super(FeatureSegmentViewSet, self).get_serializer(*args, **kwargs)
 
+    def destroy(self, request, *args, **kwargs):
+        instance = self.get_object()
+
+        # feature state <-> feature segment relationship is incorrectly modelled as a
+        # foreign key instead of one to one, so we need to grab the first feature state
+        feature_state = instance.feature_states.first()
+        message = SEGMENT_FEATURE_STATE_DELETED_MESSAGE % (
+            instance.feature.name,
+            instance.segment.name,
+        )
+        audit_log_record = (
+            AuditLog.create_record(
+                obj=feature_state,
+                obj_type=RelatedObjectType.FEATURE_STATE,
+                log_message=message,
+                author=request.user,
+                project=instance.feature.project,
+                environment=instance.environment,
+                persist=False,
+            )
+            if feature_state
+            else None
+        )
+
+        self.perform_destroy(instance)
+        response = Response(status=status.HTTP_204_NO_CONTENT)
+
+        if audit_log_record and response.status_code == status.HTTP_204_NO_CONTENT:
```

Comment:
Why are we checking the status code, here? Also, are assuming a scenario where a feature segment will exist without feature state? I think we should log an error in that case, no?

Reply:
Good question regarding the status code check - I see that that is useless 😄 I was trying to confirm that the delete was successful but I realised that if it's not, an exception would be thrown and DRF would handle returning the relevant status code. 

I've moved the logic to the `perform_destroy` method now instead which feels a bit neater. 

Regarding the check for null feature state, it's because it's entirely possible for this case to happen so I want to ensure we don't hard error. The reason it's possible is because the way the API is setup is that the FE creates a 'segment override' with 2 separate requests. One to create the FeatureSegment, then another to create the FeatureState. So if, for whatever reason, the creation of the FeatureState fails then we are left with a FeatureSegment that has no feature states.